zarr/storage.py

```diff
@@ -758,6 +758,9 @@ def __setitem__(self, key, value):
         if not os.path.exists(dir_path):
             try:
                 os.makedirs(dir_path)
+            except OSError:
+                if not os.path.isdir(dir_path):
+                    raise KeyError(key)
             except Exception:
                 raise KeyError(key)
```

Comment:
Hmm...so this `except` case doesn't catch it?

Reply:
The `except Exception` block was catching it before, but it was just throwing `KeyError` instead--so my code would still error out. The block that I added only raises `KeyError` if the folder doesn't exist already. This way, workers can go on with their business if the folder already exists.

Reply:
Should we then drop the `except Exception` entirely and only use `except OSError`? It seems bad to mask other errors by raising `KeyError` instead.

Reply:
I think that's a good idea. I wasn't sure about that part so I left it, but I'm not sure what other exceptions you can even expect from `os.makedirs`. If you got `OSError` because of permissions or something else that prevented making the folder, then the block that I added would raise `KeyError` anyway.

Reply:
Good point. We can improve this a bit then by checking the `errno` of the `OSError`. Namely if we find `e.errno == errno.EEXIST`, then we know this is a `FileExistsError`. If it's any other `errno`, we can just re-raise the exception as is.

ref: https://docs.python.org/3/library/errno.html#errno.EEXIST

Reply:
Have taken the liberty of pushing a commit to your branch using the `errno` bit. Please let me know what you think.

Reply:
> I wasn't sure about that part so I left it...

That makes sense. Honestly this is an orthogonal issue. So we need not address it here. Will raise an issue to follow-up on this after.

Reply:
Went ahead and dropped the generic exception case as it didn't seem to be covered. Happy to revisit if we figure out it was needed for some reason.